freva-client/src/freva_client/query.py

```diff
@@ -712,103 +745,132 @@ def _facet_search(
-
-        Uing this, user would be able to delete the user's data from the databrowser
-        based on the provided search keys.
-
-        Parameters
-        ~~~~~~~~~~
-        username: str
-            The username associated with the data to be deleted.
-        search_keys: dict[str, str]
-            A dictionary containing the search keys to identify the data to be deleted.
 
-        Returns
-        ~~~~~~~~
-        None
-            If the operation is successful, no return value is provided.
-
-        Raises
-        ~~~~~~~
-        ValueError
-            If the operation fails to delete the user data.
+        Deleting user data:
 
-        Example
-        ~~~~~~~
         .. execute_code::
 
-            from freva_client import databrowser, authenticate
+            from freva_client import authenticate, databrowser
             token_info = authenticate(username="janedoe")
             db = databrowser()
-            db.delete_user_data(
-                "janedoe",
-                {"project": "cmip5", "experiment": "something"}
+            db.userdata(
+                action="delete",
+                metadata={"project": "cmip5", "experiment": "myFavExp"}
             )
         """
-        url = f"{self._cfg.userdata_url}/{username}"
+        url = f"{self._cfg.userdata_url}"
         token = self._auth.check_authentication(auth_url=self._cfg.auth_url)
         headers = {"Authorization": f"Bearer {token['access_token']}"}
-        data = search_keys
-        result = self._delete(url, headers=headers, json=data)
-        if result is None:
-            raise ValueError("Failed to delete user data")
+
+        if action == "add":
+            if not userdata_items:
+                raise FileNotFoundError(
+                    "No data items found. Please provide data to add."
+                )
+
+            self.metadata_collection: List[
+                Dict[str, Union[str, List[str], Dict[str, str]]]
+            ] = []
+            self.executor = concurrent.futures.ThreadPoolExecutor(
+                max_workers=min(mp.cpu_count(), 15)
+            )
+            try:
+                self.validated_userdata = self._validate_user_data(userdata_items)
+                asyncio.run(self._process_user_data())
+            finally:
+                self.executor.shutdown(wait=True)
+                logger.info("Executor shutdown completed successfully.")
+
+            if self.user_metadata:
+                self.payload_metadata = {
+                    "user_metadata": self.user_metadata,
+                    "facets": metadata,
+                }
+                result = self._post(url, data=self.payload_metadata, headers=headers)
+                pprint("[b][green]User data added successfully[green][b]")
+                if result is None:
+                    raise ValueError("Failed to add user data")
+            else:
+                raise ValueError("No metadata generated from the input data.")
+
```

Comment:
Just to let you know, users will create instances of this object. Because of that we should be careful with attributes like user_metadata, payload_metadata etc. They aren't for the users. So anything that is a utility in this class should start with a `_` to indicate to the users that it's not for direct access. In general, I think it might be wiser to outsource most of the functionality of used in this method (validation, processing, ...) to a different user data class that is instantiated and garbage collected in this method.